### ISSUE TITLE
Simplify URLs after a search

### ISSUE THREAD

**Issue by JohnCremona** [LMFDB/lmfdb#2277]
I just did a search on cubic fields which are unramified outside {2,37}, of which there are 3 (one cyclic and 2 not).  This is beautifully simple, just go to http://www.lmfdb.org/NumberField and put 3 into the degree box and 2,37 into the ramified primes, selecting "subset of".  The result is the page whose URL displays as

http://www.lmfdb.org/NumberField/?degree=3&signature=&discriminant=&galois_group=&ur_primes=&ram_quantifier=contained&ram_primes=2%2C37&class_number=&class_group=&count=20&search=Search

which is very ugly as it contains blanks for lots of search fields I did not specify.  But I want to email that link to a colleague, and also put it in a paper!  I can manually edit the URL, deleting the irrelevant bits, to get

http://www.lmfdb.org/NumberField/?degree=3&ram_quantifier=contained&ram_primes=2%2C37

which is quite a lot better, only the encoding of the comma near the end being a bit weird.  (You can edit %2C into , and that works, at least on one browser).

Can we do better?  I can see many people who might want to show in a paper the complete list of objects in the database with some property like this, and having a more user-friendly URL would then be useful.  Is it possible to eliminate the blanks fomr the URL?

**Comment by AndrewVSutherland** [comment]
Yes, we can do better.  Indeed, the genus 2 code already does this.

**Comment by JohnCremona** [comment]
That is good to hear.  I don't have time to look into how right now, hence this issue.  Similar things would be good elsewhere for the same reason.

**Comment by JohnCremona** [comment]
@jwj61 Would I be justified in writing in a paper that there are exactly three cubic fields unramified outside {2,37} _because_ the LMFDB only has these three?  Just wondering, I do have in independent proof (it is only an example).  I assume not.

**Comment by davidfarmer** [comment]
If you go to the Jones-Roberts database and do the same search,
it says "Results below are proven complete."  Maybe some day we
will see those words in the LMFDB!

**Comment by JohnCremona** [comment]
Thanks @davidfarmer

**Comment by jwj61** [comment]
@davidfarmer is right.  I was recently thinking about porting that feature to the LMFDB.  It took me a long time the first time around.

For now, the lmfdb is enough based on its completness information.  The simple upper bound for the p-part of the discriminant a cubic ramified at 2 is 2^3 (being locally a quadratic*linear) and at a tame prime p is p^2.  Putting it together you get 10952, which is well below the discriminant bound we list for cubics on the completeness page.

**Comment by JohnCremona** [comment]
@jwj61 Is there a simple criterion for a cyclic cubic extension L/K to be unramified at a prime P dividing 3?  You can assume that the relative extension is given by a monic integral cubic f in K[x], and that the  valuation of disc(f) at P is positive and even (otherwise even I can give the answer).  Note that I do not assume that K contains cube roots of unity so f may not be a "pure cubic".

I hope that asking this here is not an abuse of github Issues.

**Comment by jwj61** [comment]
Nothing leaps to mind right off.

**Comment by JohnCremona** [comment]
It turns out that you can find a local maximal order very much faster than the global max. order (thanks to pari functionality available in Sage), so I can find the 3-maximal order of the top field L and get its absolute discriminant, then I am hope and dry if that is not divisible by 3.  Then the only remaining issue is that the base field may be ramified at 3 and its discriminant will show up in that absolute disc. so one needs to remember the formula for discriminants in towers.

**Comment by jwj61** [comment]
I thought you meant a theoretical criterion.  Yes, local maximal orders are faster, although often the main time savings is in factoring the polynomial discriminant.

**Comment by roed314** [comment]
Back to the original issue....

There are two approaches to fixing this that I found by googling.
* Add some javascript to strip out the empty fields.  See https://www.billerickson.net/code/hide-empty-fields-get-form/
* Redirect to a different URL without the empty fields.  This is a bit less hacky than the solution above, but adds an additional GET request.  See https://stackoverflow.com/questions/25335059/hide-the-text-input-field-portion-in-a-forms-get-query-url-when-the-value-is-an

Any preferences?  I don't know that I'm going to have time to do this this week, since I'm leaving for Calgary early tomorrow and teaching this afternoon.

**Comment by AndrewVSutherland** [comment]
There is already code tyo do this in our standard hompage template, you just need to bind onsubmit to cleansubmit, see g2c_browse.html, for example.

**Comment by roed314** [comment]
Thanks.  For abelian varieties, this is now https://github.com/LMFDB/lmfdb/pull/2287

**Comment by JohnCremona** [comment]
#2278 does a lot of this (thanks @jwj61 ).   I'll close it after I have done a bit more in the ec & ecnf pages concerning default parameters not needing to be set.

**Comment by AndrewVSutherland** [comment]
@JohnCremona Can this be closed or is there still more you want to do?

**Comment by JohnCremona** [comment]
I did not come back to this and do not actually understand the mechanism.  In the template ec-index.html the main search form does start with 
`<form id='search' onsubmit="cleanSubmit(this.id)">
`
but still the URL has lots of empty fields, whereas g2c_browse.html has that same line but produces nice urls.  I'll do if it someone can explain.

**Comment by AndrewVSutherland** [comment]
@JohnCremona It looks to me like the ec and ecnf browse pages are behaving exactly as they should.  For example, if I go to http://www.lmfdb.org/EllipticCurve/ and click on the search button without choosing any parameters, the URL on the browse page is:

http://www.lmfdb.org/EllipticCurve/?include_base_change=on&include_Q_curves=include&include_isogenous=on&include_cm=include

It contains no empty attributes, so cleanSubmit is doing its job.  If you want to get ride of 

- include_base_change=on
- include_Q_curves=include
- include_isogenous=on
- include_cm=include

the thing to do would be to have the html template not set these attributes (leave them blank so cleanSubmit will get rid of them) and to then have the query code apply these values as defaults (e.g. use get("include_base_change","on") when accessing the query string).  I leave it to you to decide if this is a change you think should be made, or whether it is better to have these defaults explicitly stated in the URL.  Personally I could be convinced to go either way.

**Comment by AndrewVSutherland** [comment]
@JohnCremona I realized that in fact the only change needed to get rid of the 4 attributes above is to simply change "on" to "" and "include" to "" in the appropriate places in the html templates.  I just submitted a PR that does this -- take a look at it and feel free to merge or close as you see fit.

I made a similar change to the ec search to get rid of include_cm=include.  Getting rid of surj_quanitifier=include would actually require some code changes to the query processing to make it work properly (because it is used in conjunction with the list of maximal primes) -- the PR does not make this change.
